dask/_task_spec.py

```diff
@@ -0,0 +1,928 @@
+            continue
+        new_dsk[k] = t
+    new_dsk2 = {
+        k: v if isinstance(v, BaseTask) else LiteralTask(k, v)
+        for k, v in new_dsk.items()
+    }
+    return new_dsk2
+
+
+def resolve_aliases(dsk: dict, keys: set, dependents: dict) -> dict:
+    """Remove trivial sequential alias chains
+
+    Example:
+
+        dsk = {'x': 1, 'y': Alias('x'), 'z': Alias('y')}
+
+        resolve_aliases(dsk, {'z'}, {'x': {'y'}, 'y': {'z'}}) == {'z': 1}
+
+    """
+    if not keys:
+        raise ValueError("No keys provided")
+    dsk = dict(dsk)
+    work = list(keys)
+    seen = set()
+    while work:
+        k = work.pop()
+        if k in seen or k not in dsk:
+            continue
+        seen.add(k)
+        t = dsk[k]
+        if (
+            isinstance(t, Alias)
+            and t.key not in keys
+            and t.key != k
+            and t.key in dsk
+            and len(dependents[t.key]) == 1
+        ):
+            t = dsk[k] = dsk.pop(t.key).copy()
+            seen.discard(k)
+            if isinstance(t, Alias):
+                work.append(k)
+            else:
+                t.key = k
+
+        work.extend(t.dependencies)
+    return dsk
+
+
+class KeyRef:
+    val: KeyType
+    __slots__ = ("key",)
+
+    def __init__(self, key: KeyType):
+        self.key = key
+
+    def __str__(self):
+        return str(self.key)
+
+    def __repr__(self):
+        return f"Key({repr(self.key)})"
+
+    def __eq__(self, value: object) -> bool:
+        if isinstance(value, KeyRef):
+            return self.key == value.key
+        return self.key == value
+
+    def __hash__(self) -> int:
+        return hash(self.key)
+
+
+_func_cache: MutableMapping = {}
+_func_cache_reverse: MutableMapping = {}
+
+
+class BaseTask:
+    key: KeyType
+    dependencies: set | frozenset | tuple
+
+    __slots__ = tuple(__annotations__)
+
+    def ref(self):
+        return Alias(self.key)
+
+    def copy(self):
+        raise NotImplementedError
+
+    def _verify_values(self, values: tuple | dict) -> None:
+        if not self.dependencies:
+            return
+        if missing := set(self.dependencies) - set(values):
+            raise RuntimeError(f"Not enough arguments provided: missing keys {missing}")
+
+    def inline(self, dsk) -> BaseTask:
+        raise NotImplementedError("Not implemented")
+
+    def propagate_literal(self) -> BaseTask:
+        """Return a literal if all tasks are literals"""
+        return self
+
+    def __call__(self, values) -> Any:
+        raise NotImplementedError("Not implemented")
+
+    @property
+    def is_coro(self) -> bool:
+        return False
+
+    def __sizeof__(self) -> int:
+        return sys.getsizeof(type(self)) + sizeof(self.key) + sizeof(self.dependencies)
+
+
+_no_deps: set = set()
+
+
+class Alias(BaseTask):
+    __weakref__: Any = None
+    __slots__ = tuple(__annotations__)
+
+    def __init__(self, key: KeyRef | KeyType):
```

Comment:
The `Alias` class feels slightly off. It's the only `BaseTask` class that does not return the key with which it's also registered - or at least, should be registered with - into `dsk`. Instead, it returns the aliased key. I think that the intention is to make it simple to skip the Alias, but we have to implement code for optimizing `Alias`es away nonetheless since users could also just explicitly use the key from `dsk`. This might be a pain we're facing because the existing API forces us into allowing `Alias`es. A more explicit variant would be something like `Alias(old, new)`

Reply:
An earlier version distinguished those and I had very often `Alias(key, key)` objects which felt a little redundant but I think I can make it work. It also struck me as weird to have the `key` attribute allowed to be different.

`Alias(key, ref=None)` seems appropriate

Reply:
I refactored this in https://github.com/dask/dask/pull/11248/commits/60df5d92356edec427093209b9cf45d9fafca992

I used `target` for the variable name instead of ref or new (since new is not always true, sometimes this is a genuine reference and ref is already taken)

Reply:
This also makes the `resolve_aliases` (function name is not ideal) code a little easier to read imo

Reply:
Sounds good, `target` definitely works. I wanted to avoid something ambiguous like `Alias(key, alias)` where it's unclear what the newly introduced name is.